### PR TITLE
Long symbol names for dbc export

### LINKED ISSUE #167: Long object names in DBC not fully supported
I try to rename certain objects in a dbc file (messages and signals).
The dbc supports a "normal" symbol length up to 32 chars. If longer names shall be used, that seems to be done "in the background" by special attributes, e. g. with a line like that:
`BA_ "SystemSignalLongSymbol" SG_ <frame_id> <signal name up to 32 chars> " <long signal name>";`
Reading that attributes works fine (always reports the correct name, i. e. the long name, if applicable).
Writing/Changing that attribute however fails, or let's say, it seems to work, but unfortunately "forgets" to change the special attribute.
When opening the modified dbc again, the original name is still there ...

### PR DESCRIPTION
**Pull request by geo4711**
Main change:
Closes #167: DBC export: improved handling for long symbol names (nodes, messages, signals); prepared for env_vars, too (but currently they are not exportet to dbc at all);

Minor improvements:
- dbc version string (add write access, map default value None to '')
- correct BA Enum for kind = None (as already implemented for other kinds)
- improve robustness of 1 testcase dealing with binary dbc comparision: tolerate differences caused by os specific newlines

### CODE REVIEW

**Comment by eerimoq** [comment]
Could you squash all commits into one and write the PR description as commit message?

File: tests/test_database.py [hunk 1]
```diff
@@ -4206,6 +4210,12 @@ def test_long_names_dbc(self):
         self.assertEqual(db.messages[7].signals[2].receivers,
                          ['N123456789012345678901234567890123'])
 
+        # environment variables
+        envvar_names = db.dbc.environment_variables
+        self.assertTrue('E1234567890123456789012345678901' in envvar_names)
```
Comment (eerimoq) [inline tests/test_database.py]: Use `assertIn()` and `assertNotIn()` instead.

File: tests/test_database.py [hunk 2]
```diff
@@ -4666,22 +4676,170 @@ def test_issue_163_dbc_newlines(self):
...
+
+        db_readback = cantools.database.load_string(content, 'dbc')
+
+        self.assertEqual(set([MSG_LONG, MSG_SHORT]),
+                         set([msg.name for msg in db_readback.messages]))
+        self.assertEqual(set([SIG_LONG, SIG_SHORT]),
+                         set([sig.name for msg in db_readback.messages
+                              for sig in msg.signals]))
+        self.assertEqual(set([NODE_LONG, NODE_SHORT]),
+                         set([node.name for node in db_readback.nodes]))
```
Comment (eerimoq) [inline tests/test_database.py]: Why compare dumped DBC with a reference file instead of lots of assertions?

File: tests/test_database.py [hunk 3]
```diff
@@ -4666,22 +4676,170 @@ def test_issue_163_dbc_newlines(self):
...
+        self.assertEqual(db_readback.messages[0].senders, [NODE_NAME])
+
+    def test_database_version(self):
+        # default value if db created from scratch (map None to ''):
+        db = cantools.database.Database()
+        self.assertIsNone(db.version)
+        self.assertTrue(db.as_dbc_string().startswith('VERSION ""'))
+
+        # write access to version attribute
+        MY_VERSION = "my_version"
```
Comment (eerimoq) [inline tests/test_database.py]: Local variables in lower case only. Applies everywhere.

File: tests/test_database.py [hunk 4]
```diff
@@ -4666,22 +4676,170 @@ def test_issue_163_dbc_newlines(self):
...
+            'OBJ_56789_123456789_1234567_0000' : 'OBJ_56789_123456789_1234567_0000',
+            'OBJ_56789_123456789_123456789_ABC': 'OBJ_56789_123456789_1234567_0001',
+            'OBJ_56789_123456789_123456789_ABD': 'OBJ_56789_123456789_1234567_0002',
+            },
+            {
+            'OBJ_56789_123456789_1234567_0001' : 'OBJ_56789_123456789_1234567_0001',
+            'OBJ_56789_123456789_123456789_ABC': 'OBJ_56789_123456789_1234567_0000',
+            'OBJ_56789_123456789_123456789_ABD': 'OBJ_56789_123456789_1234567_0002',
+            },
+            )
```
Comment (eerimoq) [inline tests/test_database.py]: Fix styling as:

``` python
test_vectors = (
    {
    },
    {
        'OBJ_long9_123456789_123456789_ABC' : 'OBJ_long9_123456789_123456789_AB',
        'OBJ_long9_123456789_123456789_DEF' : 'OBJ_long9_123456789_123456789_DE',
        'OBJ_short_123456789':                'OBJ_short_123456789',
        'OBJ_56789_123456789_123456789_GH' :  'OBJ_56789_123456789_1234567_0000',
        'OBJ_56789_123456789_123456789_GHI':  'OBJ_56789_123456789_1234567_0001',
        'OBJ_56789_123456789_123456789_GHIJ': 'OBJ_56789_123456789_1234567_0002'
    },
    {
        'OBJ_56789_123456789_1234567_0000' : 'OBJ_56789_123456789_1234567_0000',
        'OBJ_56789_123456789_123456789_ABC': 'OBJ_56789_123456789_1234567_0001',
        'OBJ_56789_123456789_123456789_ABD': 'OBJ_56789_123456789_1234567_0002'
    },
    {
        'OBJ_56789_123456789_1234567_0001' : 'OBJ_56789_123456789_1234567_0001',
        'OBJ_56789_123456789_123456789_ABC': 'OBJ_56789_123456789_1234567_0000',
        'OBJ_56789_123456789_123456789_ABD': 'OBJ_56789_123456789_1234567_0002'
    })
```

File: cantools/database/can/formats/dbc.py [hunk 5]
```diff
@@ -420,6 +435,59 @@ def long_to_short_names_dict(self):
...
+    exactly 32 chars.
+
+    """
+
+    def __init__(self, names):
+        self._result_dict = {}
+
+        num_objs = len(names)
+        if num_objs == 0:
+            return
```
Comment (eerimoq) [inline cantools/database/can/formats/dbc.py]: Remove this if-statement as it is not needed.

File: cantools/database/can/formats/dbc.py [hunk 6]
```diff
@@ -420,6 +435,59 @@ def long_to_short_names_dict(self):
         return self._long_to_short_names_dict
 
 
+class UniqueNamesDict(object):
```
Comment (eerimoq) [inline cantools/database/can/formats/dbc.py]: Do you have plans on affing functionality to this class? If not I suggest replacing it with a function.

File: tests/test_database.py (hunk 1 above)
Comment (geo4711) [reply tests/test_database.py]: `assertIn()` and `assertNotIn()` unfortunately are not available in python 2.x. So I've used the more general `assertTrue()` to get a higher test coverage.
Are there further reasons to prefer the more specific methods?

Comment (eerimoq) [reply tests/test_database.py]: Ohh, ok. Nevermind then.

File: tests/test_database.py (hunk 2 above)
Comment (geo4711) [reply tests/test_database.py]: It's not a real reference file. `db_readback` is assigned lik that:
```
# genarate dbc string from the db object built up from scratch:
content = db.as_dbc_string()
...
# read back that string to double check that what has been created
# is wellformed and can be imported again:
db_readback = cantools.database.load_string(content, 'dbc')
```

I'll add some comments in this test case to make it clear.

Comment (eerimoq) [reply tests/test_database.py]: I'll probably create a reference file after the merge. Just so you know.

File: tests/test_database.py (hunk 3 above)
Comment (geo4711) [reply tests/test_database.py]: Yes, you're right. I'll correct that.

File: tests/test_database.py (hunk 4 above)
Comment (geo4711) [reply tests/test_database.py]: yes, will be done

File: cantools/database/can/formats/dbc.py (hunk 5 above)
Comment (geo4711) [reply cantools/database/can/formats/dbc.py]: yes, corrected.

File: cantools/database/can/formats/dbc.py (hunk 6 above)
Comment (geo4711) [reply cantools/database/can/formats/dbc.py]: No more plans, so I'll convert it into a function.

**Comment by eerimoq** [comment]
Sure, go ahead and create a new PR.

File: tests/test_database.py (hunk 1 above)
Comment (geo4711) [reply tests/test_database.py]: I just had a look at [docs.python.org](https://docs.python.org/3/library/unittest.html?highlight=assertin#assert-methods). They say "new in 3.1"

Comment (eerimoq) [reply tests/test_database.py]: https://docs.python.org/2.7/library/unittest.html?highlight=assertin#assert-methods

Comment (geo4711) [reply tests/test_database.py]: Hmm, that's funny. I'll adapt the code and hope, that your link wins about mine ;-)

Comment (eerimoq) [reply tests/test_database.py]: =)

**Comment by geo4711** [comment]
PR rejected due to too many single commits. New one just created ...

**Pull request closed without merging**